### PR TITLE
internal: don't depend on cwd

### PR DESCRIPTION
**Pull request by matklad**
bors r+
🤖

### CODE REVIEW

**Comment by bjorn3** [comment]
Would it make sense to use cap-std in rust-analyzer everywhere instead of `std::fs`?

**Comment by matklad** [comment]
Yeah, I am brewing something along those lines:

```rust
//! Linear types can change the world!

use std::{ffi::{OsStr, OsString}, fs, io, path::{Path, PathBuf}};

use paths::{AbsPath, AbsPathBuf};

pub struct Io {
    repr: IoRepr,
}

impl Io {
    pub fn new_real() -> Io {
        let repr = IoRepr::RealIo(RealIo);
        Io { repr }
    }

    // region:process
    pub fn utf8_output(&mut self, cmd: Cmd) -> io::Result<String> {
        match &mut self.repr {
            IoRepr::RealIo(repr) => repr.utf8_output(cmd),
        }
    }
    // endregion:process

    // region:fs
    pub fn exists(&mut self, path: &AbsPath) -> bool {
        match &mut self.repr {
            IoRepr::RealIo(repr) => repr.exists(path),
        }
    }
    pub fn read_dir(&mut self, path: &AbsPath) -> io::Result<Vec<io::Result<AbsPathBuf>>> {
        match &mut self.repr {
            IoRepr::RealIo(repr) => repr.read_dir(path),
        }
    }
    // endregion:fs
}

pub struct Cmd {
    pub prog: PathBuf,
    pub current_dir: AbsPathBuf,
    pub args: Vec<OsString>,
}

```

I've looked briefly at cap-std, but it seems rather complex, and I didn't see immediatelly if it's possible to mock it out.

File: crates/paths/src/lib.rs [hunk 1]
```diff
@@ -168,6 +183,43 @@ impl AbsPath {
     pub fn strip_prefix(&self, base: &AbsPath) -> Option<&RelPath> {
         self.0.strip_prefix(base).ok().map(RelPath::new_unchecked)
     }
+    pub fn starts_with(&self, base: &AbsPath) -> bool {
+        self.0.starts_with(&base.0)
+    }
+    pub fn ends_with(&self, suffix: &RelPath) -> bool {
+        self.0.starts_with(&suffix.0)
```
Comment (lnicola) [inline crates/paths/src/lib.rs:190]: 😬

Comment (matklad) [reply crates/paths/src/lib.rs:190]: That's what happens when you add API "for completness" )

File: crates/rust-analyzer/src/reload.rs [hunk 2]
```diff
@@ -83,22 +84,22 @@ impl GlobalState {
             if path.extension().unwrap_or_default() != "rs" {
                 return false;
             }
-            if IMPLICIT_TARGET_FILES.iter().any(|it| path.ends_with(it)) {
+            if IMPLICIT_TARGET_FILES.iter().any(|it| path.as_ref().ends_with(it)) {
```
Comment (lnicola) [inline crates/rust-analyzer/src/reload.rs:87]: Any particular reason to use `as_ref()` here? It papered over the bug above.

Comment (matklad) [reply crates/rust-analyzer/src/reload.rs:87]: Was too lazy to convert constant string arrays to relative paths

Comment (lnicola) [reply crates/rust-analyzer/src/reload.rs:87]: Ah, right, that makes sense.

**Pull request merged**